internal_transfers/actions/src/accounting.ts

```diff
@@ -0,0 +1,65 @@
+import { address as SETTLEMENT_CONTRACT_ADDRESS } from "@cowprotocol/contracts/deployments/mainnet/GPv2Settlement.json";
+import { getSettlementCompetitionData } from "./orderbook";
+import { TransactionSimulator } from "./simulate/interface";
+import assert from "assert";
+import { partitionEventLogs } from "./parse";
+import { Log } from "@tenderly/actions";
+import { TokenImbalance } from "./models";
+import { aggregateTransfers, imbalanceMapDiff } from "./imbalance";
+import { ZeroAddress } from "ethers";
+
+// This represents the only required fields for InternalImbalance accounting
+// An object of this shape can be easily be constructed from any of
+// - ethers - `TransactionReceipt`,
+// - tenderly - `TransactionEvent`
+export interface MinimalTxData {
+  readonly from: string;
+  readonly hash: string;
+  readonly logs: Log[];
+}
+export async function getInternalImbalance(
+  transaction: MinimalTxData,
+  simulator: TransactionSimulator
+): Promise<TokenImbalance[]> {
+  const solverAddress = transaction.from.toLowerCase();
+  const competition = await getSettlementCompetitionData(transaction.hash);
+  if (competition === undefined) {
+    throw Error(`No competition found for ${transaction.hash}`);
+  }
+
+  // This is more of a monitoring system task!
+  assert(
+    solverAddress === competition.solver || competition.solver === ZeroAddress,
```

Comment:
The Orderbook DB naively back-filled the `solver_address` field in the solver competition table with zeros... There is a comment/todo about that in this project somewhere else (probably around `orderbook.fetchWinningSettlement`), but I could link it here. In reality, I think we should remove this assertion because its not the primary goal of this service. I just wanted to highlight.demonstrate that this would be the kind of place to alert on such a situation.

Reply:
Update - Yes this was captured by the following issue (follow link to comment it was derived from)

https://github.com/cowprotocol/solver-rewards/issues/226